cmd/rebuild-index/main.go

```diff
@@ -0,0 +1,552 @@
+	if err != nil {
+		return fmt.Errorf("cannot open old primary file: %w", err)
+	}
+	defer inFile.Close()
+
+	fi, err := inFile.Stat()
+	if err != nil {
+		return fmt.Errorf("cannot stat old primary file: %w", err)
+	}
+	inSize := fi.Size()
+	if inSize == 0 {
+		// File was already truncated to 0 size, but was not yet removed.
+		fmt.Fprintln(os.Stderr, "Primary file is empty", "file:", newPrimaryPath)
+		return nil
+	}
+
+	percentIncr := int64(1)
+	nextPercent := percentIncr
+	var count, prevCount int
+	start := time.Now()
+
+	sizeBuf := make([]byte, sizePrefixSize)
+	scratch := make([]byte, 1024)
+	var pos int64
+	for {
+		if _, err = inFile.ReadAt(sizeBuf, pos); err != nil {
+			if err == io.EOF {
+				// Finished reading entire primary.
+				break
+			}
+			return err
+		}
+		size := binary.LittleEndian.Uint32(sizeBuf)
+
+		if size&deletedBit != 0 {
+			pos += sizePrefixSize + int64(size^deletedBit)
+			continue
+		}
+
+		if int(size) > len(scratch) {
+			scratch = make([]byte, size)
+		}
+		data := scratch[:size]
+
+		if _, err = inFile.ReadAt(data, pos+sizePrefixSize); err != nil {
+			return fmt.Errorf("cannot read record data: %w", err)
+		}
+
+		var key []byte
+		if primaryType == "mh" {
+			br := bytes.NewReader(data)
+			mhr := multihash.NewReader(br)
+			h, err := mhr.ReadMultihash()
+			if err != nil {
+				return fmt.Errorf("error reading multihash from data: %w", err)
+			}
+			key = h
+		} else {
+			_, c, err := cid.CidFromBytes(data)
+			if err != nil {
+				return fmt.Errorf("error reading cid from data: %w", err)
+			}
+			key = c.Bytes()
+		}
+
+		// Get the key in primary storage
+		indexKey, err := primaryStore.IndexKey(key)
+		if err != nil {
+			return err
+		}
+
+		// The offset points to the start of the record, which is the size
+		// field. The size given in the block should be what is encoded in the
+		// record's size field.
+		blk := types.Block{Offset: types.Position(pos), Size: types.Size(size)}
+		if err = idx.Put(indexKey, blk); err != nil {
+			return err
+		}
+
+		pos += sizePrefixSize + int64(size)
+
+		count++
+
+		if count&1023 == 0 {
+			_, err := idx.Flush()
+			if err != nil {
+				return fmt.Errorf("error flushing index: %w", err)
+			}
+		}
+
+		percent := 100 * pos / inSize
+		if percent >= nextPercent {
+			elapsed := time.Since(start)
+			rate := int(float64(count-prevCount) / elapsed.Seconds())
+			start = time.Now()
+			prevCount = count
+
+			log.Printf("Indexed %d primary records, %d%% complete - %d records/sec\n", count, percent, rate)
+			for nextPercent <= percent {
+				nextPercent += percentIncr
+			}
+		}
+	}
+	idx.Close()
+	primaryStore.Close()
+
+	var finalDir string
+	if keep {
+		finalDir = newDir
+	} else {
+		log.Println("Removing", oldDir)
+		if err = os.RemoveAll(oldDir); err != nil {
+			return err
+		}
+		log.Println("Renaming", newDir, "to", oldDir)
+		if err = os.Rename(newDir, oldDir); err != nil {
+			return err
+		}
+		finalDir = oldDir
```

Comment:
Question please: for an existing index size of `n` Ti, what is the total memory and storage needed to rebuild the index to a different bit-size?

Reply:
At most 2x the current amount.  In reality it will be less since the rewritten valuestore will not have a freelist, will have all the deleted primary data removed.